<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.26">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Using alpseq with a custom reference – alpseq documentation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-587c61ba64f3a5504c4d52d930310e48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-58ea7672b0a4c2f0bfbbb6b0021227ce.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">alpseq documentation</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./run_alpseq.html"> 
<span class="menu-text">Running alpseq</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./sample_reports.html"> 
<span class="menu-text">Sample reports</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./custom_reference.html" aria-current="page"> 
<span class="menu-text">Using alpseq with a custom reference</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./alpseqR.html"> 
<span class="menu-text">alpseqR</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#building-the-database" id="toc-building-the-database" class="nav-link active" data-scroll-target="#building-the-database">Building the database</a>
  <ul class="collapse">
  <li><a href="#creating-the-v-gene-annotation-file" id="toc-creating-the-v-gene-annotation-file" class="nav-link" data-scroll-target="#creating-the-v-gene-annotation-file">Creating the V gene annotation file</a></li>
  <li><a href="#creating-the-v-gene-database" id="toc-creating-the-v-gene-database" class="nav-link" data-scroll-target="#creating-the-v-gene-database">Creating the V gene database</a></li>
  <li><a href="#creating-the-auxiliary-file" id="toc-creating-the-auxiliary-file" class="nav-link" data-scroll-target="#creating-the-auxiliary-file">Creating the auxiliary file</a></li>
  <li><a href="#creating-the-vdj-germline-databases" id="toc-creating-the-vdj-germline-databases" class="nav-link" data-scroll-target="#creating-the-vdj-germline-databases">Creating the V/D/J germline databases</a></li>
  <li><a href="#what-to-do-if-you-get-the-dreaded-blast-queryoptions-error-message" id="toc-what-to-do-if-you-get-the-dreaded-blast-queryoptions-error-message" class="nav-link" data-scroll-target="#what-to-do-if-you-get-the-dreaded-blast-queryoptions-error-message">What to do if you get the dreaded ‘BLAST query/options error’ message</a></li>
  </ul></li>
  <li><a href="#running-alpseq-with-the-custom-organismdatabase" id="toc-running-alpseq-with-the-custom-organismdatabase" class="nav-link" data-scroll-target="#running-alpseq-with-the-custom-organismdatabase">Running <em>alpseq</em> with the custom organism/database</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Using alpseq with a custom reference</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p><em>alpseq</em> comes with a built-in alpaca reference, built from IMGT sequences. This reference is used by both IgBLAST and matchbox, the two software we use for nanobody annotation. To use <em>alpseq</em> with your own custom references you just need to build your own IgBLAST database (matchbox uses one of the files in this database for its annotation process).</p>
<p>The IgBLAST documentation has a short guide <a href="https://ncbi.github.io/igblast/cook/How-to-set-up.html">here</a>, but below we will describe the process in more detail.</p>
<p>There are four steps:</p>
<ol type="1">
<li>Creating the V gene annotation file</li>
<li>Creating the V gene database</li>
<li>Creating the auxiliary file</li>
<li>Creating the database with all of the V/D/J genes that will be searched</li>
</ol>
<section id="building-the-database" class="level2">
<h2 class="anchored" data-anchor-id="building-the-database">Building the database</h2>
<p>Before you start, make sure you have the correct V/D/J germline files. This is explained in the IgBLAST documentation:</p>
<blockquote class="blockquote">
<p>To search IMGT germline sequences, you need to download them from IMGT web site (<a href="http://www.imgt.org/vquest/refseqh.html#VQUEST" class="uri">http://www.imgt.org/vquest/refseqh.html#VQUEST</a> ). You need to download all V, D and J sequences for whatever organisms you are interested in. Combine all V, all D and all J sequences, respectively, into separate files (i.e., one file for all V sequences, one for all D sequences and one file all for J sequences). After you have downloaded the sequences, invoke our utility tool edit_imgt_file.pl (in the bin directory in the IgBlast release package) to process these sequences (i.e., to change the long IMGT definition lines to germline gene names only). For example:</p>
<p><code>bin/edit_imgt_file.pl imgt_file &gt; my_seq_file</code></p>
</blockquote>
<p>Note: <em>alpseq</em> expects the files to be named like ‘imgt_{ORGANISM}_ighv’ and ‘imgt_{ORGANISM}_ighd’, etc so please follow this naming convention or you will need to manually edit the paths in the <code>igblast.nf</code> and/or <code>matchbox_annotate.nf</code> scripts.</p>
<section id="creating-the-v-gene-annotation-file" class="level3">
<h3 class="anchored" data-anchor-id="creating-the-v-gene-annotation-file">Creating the V gene annotation file</h3>
<p>This file annotates the germline V genes. It should be located at the path: <code>internal_data/alpaca/alpaca.ndm.imgt</code> (where you replace <code>alpaca</code> with the name of the organism you are planning to use).</p>
<p>It should be in a tab-separated format with the following columns:</p>
<p><code>IGHV1-18*01     1  75  76  99  100     150     151     174     175     288     VH  0</code></p>
<p><a href="https://b-t.cr/t/standardizing-the-format-of-a-germline-set/143/7">Using this forum post</a>, my best guess at what the numbers mean is:</p>
<ul>
<li><p>column 1: allele name (clean)</p></li>
<li><p>column 2: FR1 start</p></li>
<li><p>column 3: FR1 end</p></li>
<li><p>column 4: CDR1 start</p></li>
<li><p>column 5: CDR1 end</p></li>
<li><p>column 6: FR2 start</p></li>
<li><p>column 7: FR2 end</p></li>
<li><p>column 8: CDR2 start</p></li>
<li><p>column 9: CDR2 end</p></li>
<li><p>column 10: FR3 start</p></li>
<li><p>column 11: FR3 end</p></li>
<li><p>column 12: gene type (VH or VK or VL)</p></li>
<li><p>column 13: not sure? but usually seems to be zero in the human one</p></li>
</ul>
<p>Best way I can see to get these numbers is copy paste the sequence into IMGT/V-QUEST and scroll down to the bottom for the annotations. Then, manually enter them into the table.</p>
</section>
<section id="creating-the-v-gene-database" class="level3">
<h3 class="anchored" data-anchor-id="creating-the-v-gene-database">Creating the V gene database</h3>
<p>Next, you need to build a database for just the V genes. You can do this by using the <code>makeblastdb</code> tool (included in IgBLAST and BLAST software), on the cleaned V gene file as shown in the IgBLAST documentation:</p>
<pre><code>bin/makeblastdb -parse_seqids -dbtype nucl -in my_seq_file</code></pre>
</section>
<section id="creating-the-auxiliary-file" class="level3">
<h3 class="anchored" data-anchor-id="creating-the-auxiliary-file">Creating the auxiliary file</h3>
<p>The IgBLAST docs instruct us to ‘supply a file that has information such as CDR3 stop for germline J genes’. It has some that look like this (example shown is <code>optional_file/human_gl.aux</code>):</p>
<pre><code># The chain type, first coding frame start position, chain type, CDR3 stop.
#positions are 0-based

IGHJ1*01        0   JH  17
IGHJ1P*01       2   JH
IGHJ2*01        1   JH  18
IGHJ2P*01       0   JH
IGHJ3*01        1   JH  15
IGHJ3*02        1   JH  15
IGHJ3P*01       0   JH
IGHJ4*01        2   JH  13
IGHJ4*02        2   JH  13
IGHJ4*03        2   JH  13
IGHJ5*01        2   JH  16
IGHJ5*02        2   JH  16
IGHJ6*01        2   JH  28
IGHJ6*02        2   JH  28
IGHJ6*03        2   JH  28
IGHJ6*04        2   JH  28</code></pre>
<p>According to the IgBLAST documentation, the columns are</p>
<ol type="1">
<li>Allele name</li>
<li>Germline J gene coding frame start position (position is 0-based)</li>
<li>J gene type (JH, JK, JL)</li>
<li>The CDR3 end position for each sequence in the germline J sequence database</li>
</ol>
<p>We can again manually construct this using IMGT/V-QUEST. It won’t work on just the J genes by themselves, so we can construct an artificial Nb sequence to find out this information. Using human <code>IGHJ2*01</code> as an example:</p>
<p>Paste <code>IGHV1-18*02</code>, <code>IGHD1-14*01</code> and <code>IGHJ2*01</code> together and am looking to recreate this line: <code>IGHJ2*01        1    JH  18</code></p>
<p>I think that IMGT/V-QUEST is 1-based not 0-based and that’s why things aren’t quite lining up like it’s saying codon_start=2 but above it’s 1, and when you line up the CDR3 with the J gene, it ends at position 19 not 18.</p>
<p>I double checked with a couple of other J genes and this theory seems to be correct. So basically to get column 2, just look at the J region codon_start and minus 1. <strong>IMPORTANT: this only works if the sequence is in frame. so, need to check whether it is and if not then add some N’s after the D gene to bring it back into frame</strong></p>
<p>To get column 4, line up the CDR3 nucleotide sequence with the J gene and see which position it ends at (minus 1/ensure you are using zero based positions)</p>
<p>Following these steps I was able to make the auxiliary file for alpaca that is included in <em>alpseq</em>:</p>
<pre><code># The allele, first coding frame start position, chain type, CDR3 stop.
# positions are 0-based

IGHJ1*01    2   JH  13
IGHJ2*01    2   JH  19
IGHJ3*01    1   JH  15
IGHJ4*01    2   JH  13
IGHJ5*01    2   JH  16
IGHJ6*01    2   JH  13
IGHJ7*01    2   JH  19</code></pre>
<p>If you have an understanding of how to do this properly let me know, I would love to learn!</p>
</section>
<section id="creating-the-vdj-germline-databases" class="level3">
<h3 class="anchored" data-anchor-id="creating-the-vdj-germline-databases">Creating the V/D/J germline databases</h3>
<p>These should be located in the <code>databases/</code> folder. Essentially, copy your cleaned V/D/J germline reference files into here and run <code>makeblastdb</code> as for the V gene database before.</p>
</section>
<section id="what-to-do-if-you-get-the-dreaded-blast-queryoptions-error-message" class="level3">
<h3 class="anchored" data-anchor-id="what-to-do-if-you-get-the-dreaded-blast-queryoptions-error-message">What to do if you get the dreaded ‘BLAST query/options error’ message</h3>
<p>Once you’ve done these steps, try testing a random sequence with the IgBLAST command line tool. You might encounter this error message: <em>BLAST query/options error: Germline annotation database alpaca/alpaca_V could not be found in [internal_data] directory. Please refer to the BLAST+ user manual.</em></p>
<p>Some tips to fix this:</p>
<ol type="1">
<li>Make sure you didn’t forget to export the environment variables (<code>IGDATA</code>, with <strong>absolute path</strong> pointing to the folder that contains internal_data. In my case it was: <code>/stornext/General/data/user_managed/grpu_mritchie_1/kathleen/igblast/igdata/</code> and you also need <code>IGBLASTDB</code> which is the folder with all the databases you search. Mine was: <code>/stornext/General/data/user_managed/grpu_mritchie_1/kathleen/igblast/databases</code>)</li>
<li>Before making the database, try having extracted the contents of the taxonomy database (<a href="ftp://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz)" class="uri">ftp://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz)</a> into the same directory. Not sure why it works, but I found a seqanswers post about the same error message (just for a regular BLAST database) and this was suggested as a fix (the post is at <a href="#0">https://www.seqanswers.com/forum/bioinformatics/bioinformatics-aa/29829-deploying-a-local-version-of-blast-having-issues)</a></li>
<li>Once you’ve made the database, check with <code>blastdbcheck -dir .</code> (I believe this comes with the actual BLAST program, not IgBLAST) to make sure it’s all good.</li>
</ol>
<p>Following this procedure on not only the germline V gene database in internal_data but also the normal database in the IGBLASTDB path fixed the next error message I got: <em>Error: mdb_dbi_open: MDB_NOTFOUND: No matching key/data pair found</em></p>
</section>
</section>
<section id="running-alpseq-with-the-custom-organismdatabase" class="level2">
<h2 class="anchored" data-anchor-id="running-alpseq-with-the-custom-organismdatabase">Running <em>alpseq</em> with the custom organism/database</h2>
<p>Once you’ve followed these steps and confirmed it is working on the command line, you can try using the custom database with <em>alpseq</em>. You’ll need to set the following parameters: <code>igblast_databases</code> (the path to where the IgBLAST databases are) and <code>igblast_db_name</code> (the name of the custom organism).</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/kzeglinski\.github\.io\/alpseq_docs\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>