<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>alpseq documentation - Running alpseq</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">alpseq documentation</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./run_alpseq.html" aria-current="page"> 
<span class="menu-text">Running alpseq</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./sample_reports.html"> 
<span class="menu-text">Sample reports</span></a>
  </li>  
  <li class="dropdown-header">
 <span class="menu-text">alpseq_r.qmd</span></li>
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#running-the-pipeline" id="toc-running-the-pipeline" class="nav-link" data-scroll-target="#running-the-pipeline">Running the pipeline</a>
  <ul class="collapse">
  <li><a href="#parameters" id="toc-parameters" class="nav-link" data-scroll-target="#parameters">Parameters</a></li>
  <li><a href="#sample-sheet" id="toc-sample-sheet" class="nav-link" data-scroll-target="#sample-sheet">Sample sheet</a></li>
  <li><a href="#output" id="toc-output" class="nav-link" data-scroll-target="#output">Output</a></li>
  </ul></li>
  <li><a href="#other-useful-stuff" id="toc-other-useful-stuff" class="nav-link" data-scroll-target="#other-useful-stuff">Other useful stuff</a>
  <ul class="collapse">
  <li><a href="#sanger-mode" id="toc-sanger-mode" class="nav-link" data-scroll-target="#sanger-mode">Sanger mode</a></li>
  <li><a href="#igblast-vs-matchbox" id="toc-igblast-vs-matchbox" class="nav-link" data-scroll-target="#igblast-vs-matchbox">IgBLAST vs matchbox</a></li>
  </ul></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Running alpseq</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>This page will teach you how to use <em>alpseq</em> to analyse nanobody panning data! ðŸ¦™</p>
</section>
<section id="running-the-pipeline" class="level2">
<h2 class="anchored" data-anchor-id="running-the-pipeline">Running the pipeline</h2>
<ol type="1">
<li><p>Clone the repo (<code>git clone https://github.com/kzeglinski/alpseq.git</code>)</p></li>
<li><p>Edit the <code>nextflow.config</code> file to suit your environment (<a href="#config">see below</a> for more info).</p></li>
<li><p>Make sure you have nextflow installed on your computer</p></li>
<li><p>Run the pipeline using <code>nextflow run alpseq/main.nf -profile docker</code> or <code>nextflow run alpseq/main.nf -profile singularity</code> depending on which container engine you want to use</p></li>
</ol>
<p>There is also a nice guide on running nextflow pipelines for beginners <a href="https://bioinformaticsworkbook.org/dataAnalysis/nextflow/01_introductionToNextFlow.html#gsc.tab=0">here</a></p>
<section id="parameters" class="level3">
<h3 class="anchored" data-anchor-id="parameters">Parameters</h3>
<p>On the command line, you can see a summary of key parameters using <code>nextflow run alpseq/main.nf --help</code>.</p>
<p><strong>The most important parameters that you need to set every time (required arguments) are:</strong></p>
<ul>
<li><p><code>--out_dir</code> is the path where output files will be written to. If it doesnâ€™t exist yet, the pipeline will create it</p></li>
<li><p><code>--fastq_dir</code> is the path where the input fastq files are located</p></li>
<li><p><code>--sample_sheet</code> is the path to the .csv formatted sample sheet</p></li>
</ul>
<p><strong>Other parameters you might sometimes need to change are:</strong></p>
<ul>
<li><p><code>--use_igblast</code> whether to use IgBLAST for nanobody annotation or not (default: <code>true</code>)</p></li>
<li><p><code>--qc_only</code> whether to only perform quality control of the samples, and skip making a pan report. Turn this on if you are not analysing panning data with rounds (e.g.&nbsp;repertoire, or just sequencing unenriched libraries) (default: <code>false</code>)</p></li>
</ul>
<p>Some more advanced you shouldnâ€™t need to change often parameters are:</p>
<ul>
<li><p><code>--adapter_r1</code> and <code>--adapter_r2</code> the read 1 and read 2 adapters to trim off (default are for Illumina NextSeq)</p></li>
<li><p><code>--maximum_overlap</code> the maximum overlap of reads 1 and 2, for merging with FLASH. You might need to change this if you have a different sequencing construct design (default: <code>250</code>)</p></li>
<li><p><code>--igblast_databases</code> and/or <code>--mb_scripts</code> the paths to the IgBLAST databases and matchbox scripts to use for nanobody annotation</p></li>
<li><p><code>--template_dir</code> and <code>--extensions_dir</code> and <code>--quaro_base_yaml</code> the paths to the template files used to generate quarto reports. You could change these to create your own custom reports</p></li>
<li><p><code>--sequence_trim_5p</code> and <code>--sequence_trim_3p</code> are optional sequences to trim off the 5â€™ and 3â€™ when writing out nucleotide sequences. This is useful if, for example, you want to remove bits of the nanobody DNA sequence prior to writing it out (e.g.&nbsp;remove his tag) (default: <code>""</code>)</p></li>
<li><p><code>--chunk_size</code> since IgBLAST doesnâ€™t scale well, break samples into chunks for annotation. This will submit a whole bunch of slurm jobs, so you might need to change it if that is a problem for your compute cluster (default: <code>25000</code>)</p></li>
</ul>
<section id="config" class="level4">
<h4 class="anchored" data-anchor-id="config">Nextflow config files</h4>
<p>You can set all of the above parameters on the command line (e.g.&nbsp;<code>nextflow run alpseq/main.nf -profile docker --fastq_dir ~/my_files/</code>), although take care that, as for all nextflow pipelines, a single dash is used for nextflow options (e.g.&nbsp;<code>-profile docker</code>) and a double dash is used for pipeline parameters (e.g.&nbsp;<code>--fastq_dir ~/my_files/</code>).</p>
<p>However, I personally prefer to change these parameters in the <code>nextflow.config</code> file that is contained within the top-level directory of the pipeline. This is handy as you can then copy the config into your output directory to have a record of the parameters you used.</p>
<p>The config file also allows you to control other aspects of how nextflow runs this pipeline (for example, using SLURM or AWS). The config file supplied with <em>alpseq</em> has been written to run on WEHIâ€™s Milton HPC, and may not work on your system. You might want to try looking at the <a href="https://github.com/nf-core/configs">nf-core collection of institutional config files</a> to see if they have one for your institute. Alternatively, you could try asking a nextflow expert at your workplace!</p>
<p>For more information about writing/editing nextflow config files, take a look at the <a href="https://www.nextflow.io/docs/latest/config.html">official nextflow documentation</a>.</p>
</section>
</section>
<section id="sample-sheet" class="level3">
<h3 class="anchored" data-anchor-id="sample-sheet">Sample sheet</h3>
<p>Sample sheets should have the following columns:</p>
<ul>
<li><p><code>sample_num</code>: a sample number (can duplicate samples if e.g.&nbsp;the same round 0 is shared across multiple pans)</p></li>
<li><p><code>library</code>: a string to name the library (just used for naming/labelling)</p></li>
<li><p><code>antigen</code>: a name for the antigen (just used for naming/labelling)</p></li>
<li><p><code>round</code>: the round of panning e.g.&nbsp;0, 1, 2, 3, 4</p></li>
<li><p><code>replicate</code>: an ID for the replicate <code>NA</code> if no replicates</p></li>
<li><p><code>panning_id</code>: an ID that groups samples from the same pan (so that a report can contain comparisons of multiple pans)</p></li>
<li><p><code>report_id</code>: a name for the report (so you can generate multiple reports for multiple different pans at once)</p></li>
<li><p><code>r1_file_name</code>: name of the read 1 fastq file</p></li>
<li><p><code>r2_file_name</code>: name of the read 2 fastq file</p></li>
</ul>
<p>For example, the sample sheet to process the <a href="https://pubmed.ncbi.nlm.nih.gov/32887876/">Hanke et al.</a> data would be:</p>
<pre><code>sample_num,library,antigen,round,replicate,panning_id,report_id,r1_file_name,r2_file_name
S01,Tyson,covid,0,NA,1,hanke_et_al,round0_read1.fastq.gz,round0_read2.fastq.gz
S02,Tyson,covid,1,NA,1,hanke_et_al,round1_read1.fastq.gz,round1_read2.fastq.gz
S03,Tyson,covid,2,NA,1,hanke_et_al,round2_read1.fastq.gz,round2_read2.fastq.gz</code></pre>
</section>
<section id="output" class="level3">
<h3 class="anchored" data-anchor-id="output">Output</h3>
<p><em>alpseq</em> will output the following:</p>
<ul>
<li><p><code>original_annotation_tsv</code>: the .tsv files from either IgBLAST or matchbox</p></li>
<li><p><code>processed_tsv</code>: the .tsv files that have been processed by alpseq (summaries of germline gene composition, tables of clone counts collapsed by CDR3, etc)</p></li>
<li><p><code>report</code>: the QC and panning reports generated by alpseq. See the â€˜sample reportsâ€™ tab for examples of these</p></li>
</ul>
</section>
</section>
<section id="other-useful-stuff" class="level2">
<h2 class="anchored" data-anchor-id="other-useful-stuff">Other useful stuff</h2>
<section id="sanger-mode" class="level3">
<h3 class="anchored" data-anchor-id="sanger-mode">Sanger mode</h3>
<p><em>alpseq</em> can be used to annotate sanger sequences with the <code>--sanger_mode</code> parameter. This isnâ€™t really a super recommended workflow, but exists as an option for our collaborators. Use at your own risk!</p>
</section>
<section id="igblast-vs-matchbox" class="level3">
<h3 class="anchored" data-anchor-id="igblast-vs-matchbox">IgBLAST vs matchbox</h3>
<p>There are two different tools that can be used for annotating nanobodies in alpseq, <a href="https://www.ncbi.nlm.nih.gov/igblast/">IgBLAST</a> (default) and <a href="https://github.com/jakob-schuster/matchbox">matchbox</a>. A comparison/benchmark is included in our <a href="https://www.biorxiv.org/content/10.1101/2025.10.22.661623v1">preprint</a>, but briefly the tools generate highly similar counts for most CDR3s. IgBLAST is much slower, but outputs a lot more information (<a href="https://docs.airr-community.org/en/stable/datarep/rearrangements.html#fields">according to AIRR specs</a>), while matchbox is a lot faster but outputs just the bare minimum information for analysis (CDR3 and V gene). Choose the method that best suits the info you require and the computing power you have available!</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>